### PR TITLE
[Bug Fix] : writing incorrect data to target Merge repartition Command

### PR DESCRIPTION
**Pull request by paragikjain**
We were writing incorrect data to target collection in some cases of merge command. In case of repartition when source query it RELATION.  We were referring to incorrect attribute number that was resulting into this incorrect behavior.

Example :
image
image

I have added fixed tests as part of this PR , Thanks.

### CODE REVIEW

File: src/backend/distributed/planner/query_colocation_checker.c [hunk 1]
```diff
@@ -361,8 +362,9 @@ CreateAllTargetListForRelation(Oid relationId, List *requiredAttributes)
 		}
 		else
 		{
+			int varAttNum = isMergeQuery ? attrNum : colAppendIdx++;
```
Comment (onurctirtir) [inline src/backend/distributed/planner/query_colocation_checker.c:365]: I understand that there is an off-by-one error in the column values inserted into the table but I really would like understand the Postgres-internals fact that requires us doing something specific for MERGE while it's not needed for other types of commands.

Comment (paragikjain) [reply src/backend/distributed/planner/query_colocation_checker.c:365]: In Merge Command we should refer actual table attNum but in other cases we are building query like below which refers attnum of a subquery. not the actual table.
(SELECT t.a, NULL, NULL FROM (SELECT a FROM table) t).

That is the reason this is applicable only for merge command.

File: src/backend/distributed/planner/merge_planner.c [hunk 2]
```diff
@@ -857,8 +857,11 @@ ConvertRelationRTEIntoSubquery(Query *mergeQuery, RangeTblEntry *sourceRte,
 	/* set the FROM expression to the subquery */
 	newRangeTableRef->rtindex = SINGLE_RTE_INDEX;
 	sourceResultsQuery->jointree = makeFromExpr(list_make1(newRangeTableRef), NULL);
+
+	bool isMergeQuery = true;
 	sourceResultsQuery->targetList =
-		CreateAllTargetListForRelation(sourceRte->relid, requiredAttributes);
+		CreateAllTargetListForRelation(sourceRte->relid, requiredAttributes,
```
Comment (tejeswarm) [inline src/backend/distributed/planner/merge_planner.c:863]: Is this routine `CreateAllTargetListForRelation` is used by non-MERGE SQL commands?

File: src/backend/distributed/planner/merge_planner.c [hunk 3]
```diff
@@ -857,8 +857,11 @@ ConvertRelationRTEIntoSubquery(Query *mergeQuery, RangeTblEntry *sourceRte,
 	/* set the FROM expression to the subquery */
 	newRangeTableRef->rtindex = SINGLE_RTE_INDEX;
 	sourceResultsQuery->jointree = makeFromExpr(list_make1(newRangeTableRef), NULL);
+
```
Comment (tejeswarm) [inline src/backend/distributed/planner/merge_planner.c:860]: ` requiredAttributes = RequiredAttrNumbersForRelationInternal(mergeQuery, relationRestriction->index);
` 
Here were we getting 1, 3?

**Comment by tejeswarm** [comment]
Likely the problem is stemming from the fact that `ConvertRelationRTEIntoSubquery()` is really designed for

`SELECT * FROM the relation`

but in this case we are selecting column 1 and 3.  May be we should change the function to handle

`SELECT <col-list> FROM the relation`

**Comment by tejeswarm** [comment]
I think I know what's the problem is, can you please do me a small favor

Please change the below code 

```
sourceResultsQuery->targetList =
                CreateAllTargetListForRelation(sourceRte->relid, requiredAttributes);
```

to 

```
sourceResultsQuery->targetList =
                CreateFilteredTargetListForRelation(rteRelation->relid, requiredAttributes);

```

and see how things work

**Comment by paragikjain** [comment]
> ou please do me



> I think I know what the problem is, can you please do me a small favor
> 
> Please change the below code
> 
> ```
> sourceResultsQuery->targetList =
>                 CreateAllTargetListForRelation(sourceRte->relid, requiredAttributes);
> ```
> 
> to
> 
> ```
> sourceResultsQuery->targetList =
>                 CreateFilteredTargetListForRelation(rteRelation->relid, requiredAttributes);
> ```
> 
> and see how things work

Yes this works fine !! modified the changes with this function.

**Review by tejeswarm — APPROVED** [review]
LGTM, can you please add more test scenarios

In the existing test scenarios, the order of columns in both target and source are matching

INSERT (salary, id) VALUES (source.salary, source.id);
INSERT (id, salary) VALUES (source.id, source.salary);

Can we add tests where order of columns in target and source are not same i.e INSERT target(col1, col3) FROM source(col3, col1)

Semantically they may be incorrect i.e. you may be inserting age into id, but we want to make sure that the "correct columns" are picked up on both sides.

**Review by tejeswarm — APPROVED** [review]

**Pull request merged**